---
title: |
  ![](figures/mia_logo.png){width="125"} ![](figures/heart.png){width="100"} ![](figures/logo_bioconductor.gif){width="350"} \
  Microbiome data integration workflow for population cohort studies
subtitle: Package demo
output:
    html_document:
        code_folding: show
---

```{r setup, message=FALSE, class.source = 'fold-hide'}
# Get starting time
starting_time <- Sys.time()

################################################################################
# List of packages that we need
packages <- c("ggplot2", "mia", "miaViz", "dplyr", "tidyr", "scater")

# Get packages that are already installed installed
packages_already_installed <- packages[ packages %in% installed.packages() ]

# Get packages that need to be installed
packages_need_to_install <- setdiff( packages, packages_already_installed )

# Loads BiocManager into the session. Install it if it not already installed.
if( !require("BiocManager") ){
    install.packages("BiocManager")
    library("BiocManager")
}

# If there are packages that need to be installed, installs them with BiocManager
# Updates old packages.
if( length(packages_need_to_install) > 0 ) {
   install(packages_need_to_install, ask = FALSE)
}

# Load all packages into session. Stop if there are packages that were not
# successfully loaded
if( any(!sapply(packages, require, character.only = TRUE)) ){
    stop("Error in loading packages into the session.")
}

################################################################################
# Additional setup

# Set black and white theme for figures, and Arial font
theme <- theme_bw() + theme(text = element_text(family = "Arial"), 
                            panel.border = element_blank(), 
                            panel.grid.major = element_blank(),
                            panel.grid.minor = element_blank(), 
                            axis.line = element_line(colour = "black"))
theme_set(theme)
```

<img style="float: right;" src="figures/EuroBioC2023.png" width="140">

#### EuroBioC2023

- European Bioconductor Conference 2023
- Ghent, Belgium
- September 21, 2023; 13:30 (CEST)
- **See the poster also (miaverse â€“ microbiome analytics framework in SummarizedExperiment family)!**

#### Presenter information

All authors are affiliated to [Turku Data Science Group in University of Turku, Finland.](https://datascience.utu.fi/)
<img style="float: right;" src="figures/university_turku.png" width="250">

- **Tuomas Borman -- doctoral researcher**
- Chouaib Benchraka -- doctoral researcher
- Leo Lahti -- group leader

---

## Learning goals

1.  Microbiome research studies interactions between microbes (and human, environment...)
2.  Big data requires efficient tools to manipulate the data
3.  miaverse is a SummarizedExperiment framework for microbiome analytics

![Figure source: Moreno-Indias et al. (2021) Statistical and Machine Learning Techniques in Human Microbiome Studies: Contemporary Challenges and Solutions. Frontiers in Microbiology 12:11.](figures/microbiome.png){width="400"}

## Motivation

### Microbiome research

- Microbiome is a composition of microbes in well-defined area (gut, skin, mouth...)
- Bilateral interaction between human and microbiome --> affects both health and disease.
- The research is based on sequencing (characterization of genes and species).
- Nowadays, multiomics approach is more common (integration of taxonomy information with metabolite data, for example)
- Computational methods are the new microscope
- The research has expanded rapidly in previous years

```{r pubmed_fig, fig.width=6, fig.cap="PubMed publications per year with a search term 'microbiome' (fetched: Sep 5, 2023)", class.source = 'fold-hide'}
# Plot publication graph
path <- "data/PubMed_Timeline_Results_by_Year.csv"
df <- read.csv(path, skip = 1)

x <- "Year"
y <- "Count"

plot <- ggplot(df, aes(x = .data[[x]], y = .data[[y]])) +
    geom_bar(stat="identity")
plot
```

### Big data

- Most of the cohort studies require multiple types of data, for instance microbiome, phenotype and other omics data.
- Linking these types of data could present multiple challenges; to mention:
  * Different sources the data is coming form, could use different types of key identifier of the samples or subjects of the study; which could poses a problem in some cases for linking the data together.
  * The choice of the data structure to combine and store these mulitple types of data is critical as well, in terms of the management, handling and wrangling of the data; to be able to perform the analysis.
- Cohort datasets are large in size, which mostly one cannot conduct an analysis using a normal computer with limited hardware; HPC and cloud computing services are required in this case. 
- Last but not least, a quality check reporting of the constructed object is substantially important to verify the correctness of the building process.   

Regarding the choice of the data strcuture to handle cohorts with muliple data types or sources [MultiAssayExperiment](https://bioconductor.org/packages/release/bioc/html/MultiAssayExperiment.html),
is considered to be very liable in such situation, easing-up the management and wrangling of the data; once is construced correctly. Moreover, several R packages frameworks are increasingly integrating 
[MultiAssayExperiment](https://bioconductor.org/packages/release/bioc/html/MultiAssayExperiment.html) and the [SummarizedExperiment](https://bioconductor.org/packages/release/bioc/html/SummarizedExperiment.html)class, to provide users with a reliable and ease of use data structure.

## miaverse (MIcrobiome Analysis) ![](figures/mia_logo.png){width="50"}

- A framework for microbiome analytics
    - [mia (analysis)](https://bioconductor.org/packages/release/bioc/html/mia.html)
    - [miaViz (visualization)](https://bioconductor.org/packages/release/bioc/html/miaViz.html)
    - [miaSim (simulation)](https://bioconductor.org/packages/release/bioc/html/miaSim.html)
    - [Orchestrating Microbiome Analysis, OMA (tutorial book)](https://microbiome.github.io/OMA/)
- Based on _TreeSummarizedExperiment (TreeSE)_ class
    - Supports hierarchical data
    - _phyloseq_ class is a subset of _TreeSE_
    - Extension of _SingleCellExperiment_ class
    - miaverse is compatible with other _SummarizedExperiment_ frameworks
- Purpose to offer tools and tips for microbiome data analysis
- Enables development of versatile analytical workflows in microbiome data science
    - Supports multiomics (_MultiAssayExperiment_ class)
    - Scalable
    - Standardized

![The structure of the TreeSummarizedExperiment (TreeSE) class.](figures/treese.png){width="700"}

## The workflow

### Importing the dataset

We get the data from [MGnify database](https://www.ebi.ac.uk/metagenomics). It
is a EMBL-EBI's database for metagenomic data. This large microbiome database
can be accessed with _MGnifyR_ package which nowadays support _TreeSE_. The
package will be submitted to Bioconductor's next release.

We chose this dataset...

As loading takes some time, the dataset is already loaded.

For other available datasets and importing methods, check [OMA](https://microbiome.github.io/OMA/containers.html#example-data).

```{r}
# library(MGnifyR)
# mg <- MgnifyClient()
# 
# analyses <- searchAnalysis(mg, "studies", "MGYS00005128")
# analyses <- searchAnalysis(mg, "studies", "MGYS00000596")
```

```{r}
# mae <- getResult(mg, analyses)
```

```{r}
library(mia)
data("HintikkaXOData")
MAE <- HintikkaXOData
MAE
```

As we can see, our data contains three different types of data: microbiota,
metabolites and biomarkers; which are all of the TreeSummarizedExperiment class
combined in ourMultiAssayExperiment object.

Microbiota:

```{r}
MAE[["microbiota"]] # or MAE[[1]]
```

Metabolites:

```{r}
MAE[["metabolites"]] # or MAE[[2]]
```

Biomarkers:

```{r}
MAE[["biomarkers"]] # or MAE[[3]]
```

Additionally, the MAE contains phenotype or metadata about samples or subjects,
that could be accessed as follows:

```{r}
colData(MAE)
```

### Exploring the microbiome data:

#### Taxa Counts

```{r}
library(dplyr)
rowData(MAE[[1]]) %>% as_tibble() %>% summarise_all(n_distinct)
```

#### Top 10 Species

```{r}
library(miaViz)
# we agglomerate the data from OTUs to species,
# and store it at the alternative experiment slot 
altExp(MAE[[1]], "Species") <- agglomerateByRank(MAE[[1]], rank="Species",
                                                 onRankOnly=TRUE, na.rm = FALSE)
# we transform data from counts to relative abundance
altExp(MAE[[1]], "Species") <- transformAssay(altExp(MAE[[1]], "Species"),
                                          assay.type = "counts",
                                          method = "relabundance")
plotAbundanceDensity(altExp(MAE[[1]], "Species"), assay.type = "relabundance",
                     n=10L, point_size=0.5, layout="density") +
    scale_x_log10() +
    labs(title = "Top 10 Species", x="log10(relabundance) ")+
    theme(plot.title = element_text(hjust = 0.5, size = 12),
          text=element_text(size = 8))
```

#### Top 10 Genera

```{r}
# we agglomerate the data from OTUs to species,
# and store it at the alternative experiment slot 
altExp(MAE[[1]], "Genus") <- agglomerateByRank(MAE[[1]], rank="Genus",
                                                 onRankOnly=TRUE, na.rm = FALSE)
# we transform data from counts to relative abundance
altExp(MAE[[1]], "Genus") <- transformAssay(altExp(MAE[[1]], "Genus"),
                                          assay.type = "counts",
                                          method = "relabundance")
plotAbundanceDensity(altExp(MAE[[1]], "Genus"), assay.type = "relabundance",
                     n=10L, point_size=0.5, layout="density") +
    scale_x_log10() +
    labs(title = "Top 10 Genus", x="log10(relabundance) ")+
    theme(plot.title = element_text(hjust = 0.5, size = 12),
          text=element_text(size = 8))
```

#### Alpha diveristy

Lets look at ShannonIindex and Observed Richness, where the results are stored 
back automatically at the `colData` of the MAE microbiome Species alternative 
experiment:

```{r}
altExp(MAE[[1]], "Species") <- estimateDiversity(altExp(MAE[[1]], "Species"),
                                                 index="shannon", 
                                                 assay.type="counts")
altExp(MAE[[1]], "Species") <- estimateRichness(altExp(MAE[[1]], "Species"),
                                                index="observed",
                                                assay.type="counts")
colData(altExp(MAE[[1]], "Species"))
```

Let's look at the histogram of both indices:

```{r}
colData(altExp(MAE[[1]], "Species")) %>% as_tibble() %>%
    tidyr::pivot_longer(cols=c(shannon,observed), names_to = "alpha_type",
                        values_to = "estimate") %>% 
    ggplot(aes(x=estimate)) +
    geom_histogram(color = "black", fill = "gray", bins = 30)+
    facet_wrap(~alpha_type, nrow = 1, scales = "free")+
    labs(title = "Alpha Diversity")+
    theme(plot.title = element_text(hjust = 0.5))
```

#### Beta diveristy

Let's look at beta diversity by performing Principal Coordinate Analysis (PCoA)
using Bray-Curtis as a distance method on the OTU level data:

```{r}
library(scater)
MAE[[1]] <- runMDS(MAE[[1]], FUN = vegan::vegdist, name = "PCoA_BC",
              assay.type = "counts")
# Calculating explained variance
e <- attr(reducedDim(MAE[[1]], "PCoA_BC"), "eig");
rel_eig <- e/sum(e[e>0])
# Ploting the ordination 
plotReducedDim(MAE[[1]], "PCoA_BC") + 
    labs(x = paste("PC1 (", round(100 * rel_eig[[1]],1), "%", ")", sep = ""),
         y = paste("PC2 (", round(100 * rel_eig[[2]],1), "%", ")", sep = ""),
         title="PCoA with Bray-Curtis")+
    theme(plot.title = element_text(hjust = 0.5))
```

With could visualize the same plot using a metadata varible (from colData), to
overlay it on our PCoA plot:

```{r}
# we first assign our colData from MAE to MAE [[1]]
colData(MAE[[1]]) <- colData(MAE)[colnames(MAE[[1]]),]
plotReducedDim(MAE[[1]], "PCoA_BC", colour_by="Fat") + 
    labs(x = paste("PC1 (", round(100 * rel_eig[[1]],1), "%", ")", sep = ""),
         y = paste("PC2 (", round(100 * rel_eig[[2]],1), "%", ")", sep = ""),
         title="PCoA with Bray-Curtis")+
    theme(plot.title = element_text(hjust = 0.5))
```


### Thanks!

<img style="float: right;" src="figures/mia_logo.png" width="140">

- more material in OMA
- miaverse logo
- project website and QR also.
- Contact info
- Poster info



## Session info

```{r sessioninfo, class.source = 'fold-hide'}
sessionInfo()
```
