---
title: "Import data"
format:
  revealjs:
    self-contained: true
    slide-number: true
    preview-links: auto
    logo: images/mia_logo.png
    footer: <https://microbiome.github.io/>
date: last-modified
date-format: full
bibliography: references.bib
---

##

![](images/workflow.png){fig-alt="Microbiome data science workflow" fig-align="center" width=40%}

##

![](images/workflow.png){style="width:140%; height:650px; object-fit:cover; object-position:top;" fig-alt="Microbiome data science workflow" fig-align="center"}

##

![](images/workflow.png){style="width:120%; height:800px; object-fit:cover; object-position:bottom;"}

::: {.notes}

mia is for downstream analysis

:::

## Importers and converters

- `mia` includes importers and converters for standard data formats
  - Importer: Import file
  - Converter: Convert R object to other format
- You can also create `TreeSE` object manually

::: {.notes}

Importers and converters for standard data formats.

You can also create TreeSE object manually by hand if you have csv or excel files.

:::

## Open databases {.smaller}

| Data resource                                        | Package                       | # studies or datasets | # samples |
|-------------------------------------------------|------------------------------------------|-----------------------|-----------|
| Curated metagenomic data [@pasolli2017] | curatedMetagenomicData [@pasolli2017]    | 93                    | 22,588    |
| HoloFood [@rogers2024]                          | HoloFoodR [@holofoodr]                   |                      - | 9,990     |
| MGnify [@gurbich2023]                           | MGnifyR [@mgnifyr]                       | 5129                  | 616,138   |
| Microbiome benchmark data [@gamboa-tuz2025]        | MicrobiomeBenchmarkData [@gamboa-tuz2025] | 6                     | 1,125    |
| microbiomeDataSets [@microbiomedatasets]        | microbiomeDataSets [@microbiomedatasets] | 6                     | 19,100    |


::: {.notes}

There are many freely available databases for microbiome data.

For method development, benchmarking, or gaining new insights.

:::


## Demonstration

## {auto-animate="true"}

```r
library(mia)

tse <- importBIOM("file_path.biom")
print(tse)
```

```{r}
#| label: show_treese
library(mia)
data(GlobalPatterns)
tse <- GlobalPatterns
tse
```

## Exercises

From [OMA online book, Chapter 4: Import](https://microbiome.github.io/OMA/docs/devel/pages/import.html)

- Exercise 1

## References
