---
title: "Orchestrating microbiome multi-omics"
title-block-banner: true
subtitle: "Bioconductor conference, 2023"
author: "Leo Lahti @antagomir"
format: 
  revealjs:
    theme: default
    chalkboard: true
editor: visual
---

## Orchestrating microbiome multi-omics with R and Bioconductor {.smaller}

```{=html}
<!--

{background-color="white"}

To print into PDF, open the revealjs html locally and insert ?print-pdf into the url:
lahti_bioc2023.html?print-pdf#/title-slide

Then print with Margins "none"

Canvas:
https://quarto.org/docs/presentations/revealjs/presenting.html

-->
```
**Leo Lahti \@antagomir** \| Department of Computing, University of Turku, Finland

![](images/Screenshot%20from%202023-07-27%2017-47-04.png){width="29%"}

## Organizing microbiome data {.smaller}

::: r-hstack
::: columns
::: {.column width="40%"}
**Side information**\
sample & feature metadata

**Multiple assays**\
counts, relabundance, phILR..

**Alternative experiments** \
16S, metagenome, phylogenetic microarrays; pipelines

**Cross-kingdom analyses**  bacteriome, archaeome, virome, mykobiome, eukaryome

**Hierarchical data**\
taxonomic levels, nested designs, sample & feature trees
:::

::: {.column width="60%"}
![](images/Screenshot%20from%202023-07-27%2017-54-40.png)
:::
:::
:::

##  {.smaller}

<!-- Analysis workflow with *mia* {.smaller} -->

![Fig. by Domenick Braccia (EuroBioC 2020)](images/Screenshot%20from%202023-07-31%2000-25-49.png)

## Integrating multiple experiments {.smaller}

Combine **taxonomic, metagenomic & functional profiling, and host measurements** 

by extending the Bioconductor *MultiAssayExperiment* class.

::: {layout="[50, 50]"}
![](images/Screenshot%20from%202023-07-30%2015-12-18.png){height="400"}

::: {layout="[[100], [100]]"}
![](images/Screenshot%20from%202023-07-31%2000-08-19.png) ![](images/Screenshot%20from%202023-07-31%2000-14-46.png)
:::
:::

<!--$\rightarrow$ Interoperability, sustainability, efficiency-->

## Benefits for microbiome data science {.smaller}

::: columns
::: {.column width="60%"}
![](images/Screenshot%20from%202023-07-30%2015-12-18.png)
:::

::: {.column width="40%"}
**Extended**\
enhances multi-assay analyses in microbiome studies

**Optimized**\
speed, memory & scale (sample sizes, datasets, cohorts)

**Integrated**\
with other applications & frameworks (e.g. Single Cell, Spatial transcriptomics)
:::
:::

<!--$\rightarrow$ Interoperability, sustainability, efficiency-->

## Package and data ecosystem {.smaller}

Bioconductor methods & packages supporting the framework.

::: columns
::: {.column width="50%"}
![](images/Screenshot%20from%202023-07-31%2021-45-05.png)
:::

::: {.column width="50%"}
-   Basic wrangling: transformations, agglomeration, split/merge etc.

-   Alpha & beta diversity

-   Dimension reduction (scater)

-   Differential abundance (ANCOMBC, ALDEx2, benchdamic)

-   Visualization (miaViz)

-   Time series manipulation (miaTime)

-   Ecological simulations (miaSim)

![](images/mia_logo.png){width="80"} ![](images/bioconductor_logo_rgb.jpg){.absolute width="200" top="585" left="680"}
:::
:::

##  {.smaller}

### Importers for other common formats

Linking with other common formats and microbiome data science frameworks.

![](images/Screenshot%20from%202023-07-31%2022-04-20.png){height="120"} ![](images/Screenshot%20from%202023-07-31%2022-04-41.png){height="120"} ![](images/Screenshot%20from%202023-07-31%2022-03-35.png){height="120"} ![](images/Screenshot%20from%202023-07-31%2022-12-25.png){height="120"}

### Open data resources

Open microbiome data sets readily available in *TreeSummarizedExperiment* format via *curatedMetagenomicData*, *EBI MGnify*, Bioconductor *ExperimentHub*, and packages.

::: {layout="[[40, 30, 33]]"}
![](images/Screenshot%20from%202023-07-31%2022-00-40.png){width="350"}

![](images/Screenshot%20from%202023-07-31%2022-01-31.png){width="303"}

![](images/Screenshot%20from%202023-07-31%2022-02-16.png){width="420"}
:::

## Minimal example {.smaller}

Cross-correlating taxonomic and metabolomic profiles from a dietary intervention study in mice (Hintikka et al. 2021).

```{r}
#| echo: true
#| fig-width: 15
#| fig-height: 8
#| fig-pos: r
#| code-line-numbers: "false"
#| cache: true
library(mia)

# Import data and rename
data("HintikkaXOData")
mae <- HintikkaXOData
  
# Agglomerate rare taxa by prevalence at the Family level & log10-transform read  counts
mae[["microbiota"]] <- agglomerateByPrevalence(mae[["microbiota"]], rank = "Family")
mae[["microbiota"]] <- transformAssay(mae[["microbiota"]], method = "clr", pseudocount = 1)

# Get cross-correlation between taxa & metabolites
x <- testExperimentCrossAssociation(mae, experiment1 = "microbiota", experiment2 = "metabolites",
                                               assay.type1 = "clr", assay.type2 = "nmr",
                                               mode = "matrix", sort = TRUE)$cor

# Visualize taxa-metabolite associations on heatmap
ComplexHeatmap::Heatmap(x)
```

## Documentation: *OMA* Gitbook (beta) {.smaller}

Complements other Bioconductor Gitbooks; overlapping methods and analysis strategies: [microbiome.github.io/OMA]{.underline}

::: columns
::: {.column width="65%"}
![](images/Screenshot%20from%202023-07-30%2019-15-10.png)
:::

::: {.column width="35%"}
![](images/Screenshot%20from%202023-07-29%2017-13-06.png){width="350"}\
![](images/Screenshot%20from%202023-07-29%2017-11-27-01.png){width="350"}
:::
:::

## Community {.smaller}

::: columns
::: {.column width="70%"}
**Coordination:** Tuomas Borman, Leo Lahti

Giulio Benedetti, Yağmur Şimşek, Basil Courbayre, Jeba Akewak, Daena Rys, Henrik Eckermann, Chouaib Benchraka, Rajesh Shigdel, Artur Sannikov, Lu Yang, Renuka Potbhare, S. A. Shetty, R. Huang, F. G.M. Ernst, D. J. Braccia, H. C. Bravo; 20+ contributors, 10+ countries. Welcome to join!

**Acknowledgments:** TreeSummarizedExperiment, SingleCellExperiment, MultiAssayExperiment, curatedMetagenomicData, MGnifyR, phILR, ANCOMBC, ALDEx2, benchdamic, scater, scuttle authors..

**Courses in 2023:** Pune (IND) \| Turku & Oulu (FIN) \| Utrecht, Wageningen & Radboud (NLD) \| Norwich (UK) \|Bergen (NO)
:::

::: {.column width="30%"}
![](images/Screenshot%20from%202023-07-27%2018-03-25.png){.absolute width="300" top="0" left="720"}\
![](images/renuka.png){.absolute width="180" top="265" left="760"}\
![](images/ccby.png){.absolute width="120" top="535" left="800"}
:::
:::

::: columns
::: {.column width="14%"}
![](images/turun_yliopiston_logo.jpg)
:::

::: {.column width="14%"}
![](images/aka.png)
:::

::: {.column width="14%"}
![](images/biocityturku_bakgrund.jpg)
:::

::: {.column width="8%"}
![](images/cropped-Turun_Yliopistosaatio_200px.jpg)
:::

::: {.column width="14%"}
![](images/FindingPeno2.jpeg)
:::

::: {.column width="14%"}
![](images/ml4microbiome_logo.png)
:::

::: {.column width="14%"}
![](images/bioconductor_logo_cmyk.jpg)
:::
:::

## 

## Bioconductor books

::: columns
::: {.column width="50%"}
![](images/Screenshot%20from%202023-07-29%2017-13-06.png){width="350"}\
![](images/Screenshot%20from%202023-07-29%2017-11-27-01.png){width="350"}
:::

::: {.column width="50%"}
![](images/Screenshot%20from%202023-07-29%2017-48-58.png)
:::
:::

## Resources

Microbiome R package listing:

[github.com/microsud/Tools-Microbiome-Analysis](https://github.com/microsud/Tools-Microbiome-Analysis "Microbiome R packages")

::: columns
::: {.column width="20%"}
![](images/mia_logo.png)
:::

::: {.column width="80%"}
![](images/Screenshot%20from%202023-07-27%2018-02-19.png)
:::
:::

<!--Link to Shetty pkg listing-->

<!--![](images/Screenshot%20from%202023-07-27%2018-00-45.png)-->

## Microbiome data science workflow {.smaller}

::: columns
::: {.column width="32%"}
Data containers:

-   *TreeSummarizedExperiment*

-   *MultiAssayExperiment*

R/Bioc packages:\
mia, miaViz
:::

::: {.column width="68%"}
![Figure by Domenick Braccia (EuroBioC 2020)](images/Screenshot%20from%202023-07-27%2017-58-09.png)
:::
:::

## Misc

![Figure source: Moreno-Indias et al. (2021) Statistical and Machine Learning Techniques in Human Microbiome Studies: Contemporary Challenges and Solutions. Frontiers in Microbiology 12:11.](images/oma.png){width="242"}

## TreeSummarizedExperiment

![](images/Screenshot%20from%202023-07-27%2017-54-40.png)

## Optimal container for microbiome data integration? {.smaller}

Improve interoperability, ensure sustainability.

::: columns
::: {.column width="40%"}
**Multiple assays**\
seamless interlinking

**Side information**\
extended capabilities & data types

**Hierarchical data**\
both samples & features

**Optimized**\
speed & memory

**Integrated**\
other applications & frameworks
:::

::: {.column width="60%"}
![](images/ecosystem.png)
:::
:::

## Data integration in microbiomics {.smaller}

-   Scaling up sample size (cohorts & meta-analyses)

-   Complementary pipelines

-   Alternative experiments (e.g. 16S, metagenome, phylogenetic microarrays)

-   Multiple assays (e.g. counts, relative abundances, clr, phILR)

-   Hierarchies in feature & sample space (phylogenies, nested designs)

-   Cross-kingdom analyses (bacteriome, virome, mykobiome, eukaryome)

-   Multi-omics (taxonomy, function, host)

## Optimal container for microbiome data integration? {.smaller}

Improve interoperability, ensure sustainability.

::: columns
::: {.column width="40%"}
**Multiple assays**\
seamless interlinking

**Side information**\
extended capabilities & data types

**Hierarchical data**\
both samples & features

**Optimized**\
speed & memory

**Integrated**\
other applications & frameworks
:::

::: {.column width="60%"}
![](images/ecosystem.png)
:::
:::

## Linking multiple omics {.smaller}

Taxonomic, metagenomic & functional profiling, host measurements

![](images/Screenshot%20from%202023-07-30%2015-12-18.png)

## Open data resources {.smaller}

-   curatedMetagenomicData (Pasolli et al. 2017)

-   EBI MGnify (Mitchell et al. 2020)

-   microbiomeDataSets (Ernst et al. 2020);

-   R package demo data sets

-   Import functions for common formats (Biom, QIIME2, Mothur, phyloseq)
